pkg/server/api_v2_grants_test.go

```diff
@@ -0,0 +1,488 @@
+				{Grantee: "user2", Privilege: "BACKUP"},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var queryParams string
+			if tc.pageSize > 0 {
+				queryParams += fmt.Sprintf("%s=%d", pageSizeKey, tc.pageSize)
+			}
+			if tc.pageNum > 0 {
+				if len(queryParams) > 0 {
+					queryParams += "&"
+				}
+				queryParams += fmt.Sprintf("%s=%d", pageNumKey, tc.pageNum)
+			}
+			url := fmt.Sprintf("%s/api/v2/grants/databases/%d/?%s", s.AdminURL(), tc.dbID, queryParams)
+			req, err := http.NewRequest("GET", url, nil)
+			require.NoError(t, err)
+
+			clients := []*http.Client{&adminClient, &authenticatedClient}
+			if tc.client != nil {
+				clients = []*http.Client{tc.client}
+			}
+
+			for _, client := range clients {
+				resp, err := client.Do(req)
+				require.NoError(t, err)
+				defer resp.Body.Close()
+
+				require.Equal(t, tc.expectedStatus, resp.StatusCode)
+
+				if tc.expectedStatus == http.StatusOK {
+					var apiResp databaseGrantsResponseWithPagination
+					err = json.NewDecoder(resp.Body).Decode(&apiResp)
+					require.NoError(t, err)
+
+					require.Equal(t, tc.pageSize, apiResp.PaginationInfo.PageSize)
+					require.Equal(t, tc.pageNum, apiResp.PaginationInfo.PageNum)
+					require.Equal(t, tc.expectedTotal, int(apiResp.PaginationInfo.TotalResults))
+					require.Equal(t, tc.pageSize, int(apiResp.PaginationInfo.PageSize))
+					require.Equal(t, tc.pageNum, int(apiResp.PaginationInfo.PageNum))
+					require.Len(t, apiResp.Results, len(tc.expected))
+
+					for i, grant := range apiResp.Results {
+						require.Equal(t, tc.expected[i].Grantee, grant.Grantee)
+						require.Equal(t, tc.expected[i].Privilege, grant.Privilege)
+					}
+				}
+			}
+		})
+	}
+
+	// Test for non-existent database.
+	t.Run("Non-existent database", func(t *testing.T) {
+		url := fmt.Sprintf("%s/api/v2/grants/databases/99999/?limit=10&offset=0", s.AdminURL())
+		req, err := http.NewRequest("GET", url, nil)
+		require.NoError(t, err)
+		resp, err := adminClient.Do(req)
+		require.NoError(t, err)
+		defer resp.Body.Close()
+		require.Equal(t, http.StatusNotFound, resp.StatusCode)
+	})
+
+	t.Run("Retrieving grants for databases with special names", func(t *testing.T) {
+		databases := []string{
+			"mixedCaseDatabase", "database with spaces", "database-with-dashes", "databases.with spe. cial characters",
+		}
+		for _, db := range databases {
+			conn.Exec(t, fmt.Sprintf(`CREATE DATABASE "%s"`, db))
+			dbID := 0
+			conn.QueryRow(t, `SELECT id FROM system.namespace WHERE name = $1`, db).Scan(&dbID)
+
+			url := fmt.Sprintf("%s/api/v2/grants/databases/%d/?limit=10&offset=0", s.AdminURL(), dbID)
+			req, err := http.NewRequest("GET", url, nil)
+			require.NoError(t, err)
+			resp, err := adminClient.Do(req)
+			require.NoError(t, err)
+			defer resp.Body.Close()
+			require.Equal(t, http.StatusOK, resp.StatusCode)
+		}
+	})
+
+	t.Run("test bad request", func(t *testing.T) {
```

Comment:
related to the regex comment above, I would think that this is returning a 404 not found instead of the expected bad request. Am I wrong about this?

Reply:
Yeah that's right - moved the non positive integer cases to the 404 test case!